memkind hmat example (#3)
pmemhackathon/2021-04-21: examples/I/README.txt

Allocation tiering based on memory-characteristics.

Starting from Linux Kernel version 5.5, the NUMA nodes concept was enhanced with
new memory attributes: latency, capacity and bandwidth. We could use them to

> allocate data used in different scenarios on different types of memory. Note, that here

Reviewer: Note, that ~~*here*~~ we don't explicitly say which kind of medium (e.g. DRAM or PMEM) we want to use **here**

> allocate data used in different scenarios on different types of memory. Note, that here

Reviewer: Note, that ~~**here**~~ we don't explicitly say which kind of medium (e.g. DRAM or PMEM) we want to use **here**

Author: done

we don't explicitly say which kind of medium (e.g. DRAM or PMEM) we want to use, because
some characteristics (like latency or bandwidth) could be relative to relationship between
source and destination NUMA nodes. Instead of that we use a concept of "lowest latency" or
"highest bandwidth" nodes that could be accessed from a CPU that runs our program.

> This concept was added to Memkind 1.11. To use is, a hwloc library must be installed

Reviewer: ```suggestion
This concept was added to Memkind 1.11. To use it, a hwloc library must be installed
```

Author: done

on the system with kernel >= 5.5. For more information about memory attributes,
see great https://pmem.io/2021/05/12/hmat_memkind.html article on PMEM.io blog.

In this example we use  MEMKIND_LOWEST_LATENCY_LOCAL and MEMKIND_HIGHEST_CAPACITY
memory kinds to allocate simple array of data and access it through a hashmap.
Because all hashmap operations like searching or adding new data should be efficient,
keys that identifies user data entries are allocated in lowest latency memory.
On the other hand, because size of the data entires could be huge, to allocate
them we want a memory with highest capacity.

This example is intended for C programmers.

> This is example consists of these files:

Reviewer: ```suggestion
This example consists of these files:
```

possibly, to avoid duplication:

```suggestion
This example consists of the following files:
```

Author: done

hashmap.c/h -- an example of hashmap that uses custom data allocator
hmat.c      -- main file that uses Memkind and hashmap
Makefile    -- rules for building this example
run.sh      -- one way to run this example to illustrate what it does

To build this example run: make
To run it and see what it illustrates run: ./run.sh

Modifying the code and run steps is a great way to learn from this example.

> No persistence in this example, so no concerns about flushing to persistent
> or transactions for failure atomicity, etc.

Reviewer: ```suggestion
There is no persistent storage in this example, so topics such as:
 - flushing to persistent storage,
 - transactions and failure atomicity
 are not considered.
 ```

I'm not sure about persistence and persistent storage; the word "persistence" can have multiple meanings and I'm afraid that recipients of this tutorial, who might be new to pmem, might not associate it correctly with pmem functionality.

Author: Done. Note that there are other tutorials in this repo, including one that explains the basics of PMEM. Also I guess that people who attend the IntelON conference will receive materials that explain all PMEM-related questions.
